src/mpl/src/atomic/mpl_atomic.c

```diff
@@ -0,0 +1,36 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+/* FIXME For now we rely on pthreads for our IPC locks.  This is fairly
+ * portable, although it is obviously not 100% portable.  We need to figure out
+ * how to support other threading packages and lock implementations, such as the
+ * BG/P lockbox. */
+
+#ifdef MPL_HAVE_PTHREAD_H
+#include <pthread.h>
+#include "mpl_atomic.h"
+
+pthread_mutex_t *MPL_emulation_lock = NULL;
+
+int MPL_atomic_interprocess_lock_init(MPL_atomic_emulation_ipl_t * shm_lock, int isLeader)
```

Comment:
Interprocess mutex functionality was recently added to MPL in a73d3fd45437c4c4672a24ca986ba1e11b688f83. Is this redundant?

Reply:
This `lock_init` is not for users, but for OPA itself. This mutex (`OPA_emulation_lock`) is used only for lock-based atomic primitives (a fallback atomic wrapper) (https://github.com/pmodels/mpich/blob/master/src/openpa/src/primitives/opa_by_lock.h). Users cannot use it (i.e., no actual lock functions are exposed).

First of all, I don't know whether inter-process atomicity is necessary for Pthreads-based locking. Does OPA assume the following?
```
int *data = mmap(); // inter-process mmap.
void process_0_thread():
  atomic_fetch_add(&data[0]);
void process_1_thread():
  atomic_fetch_add(&data[0]);
/*
 * If atomic_fetch_add is mapped to atomic-primitives (e.g., GCC's __atomic), the update
 * to data[0] is automatically synchronized. The inter-process property is necessary with
 * Pthreads's mutex to mimic this behavior
 */
```
I don't think any atomic primitive guarantees the inter-process synchronization (at least MPL/atomic does not explicitly guarantee it). Is there a use case above in MPICH?

Reply:
I think you addressed my question of whether this is user-visible and therefore redundant. Someone else will have to answer your other questions :).

Reply:
FWIW, if this is not externally visible, should have an `MPLI_` prefix?

Reply:
Sorry for confusing you. The *functionality* of this interprocess mutex is not for users, but for OPA itself.

The trickiest part of this lock is that users cannot use this `emulation_lock`, but users explicitly need to initialize it somewhere before using OPA if Pthreads-based atomic is enabled: (e.g., https://github.com/pmodels/mpich/blob/master/src/mpid/common/shm/mpidu_shm_alloc.c#L200).
That's why this mutex initialization function is external but no one can use it.